core/services/ocr2/plugins/ocr2keeper/evmregistry/v21/logprovider/buffer_v1.go

```diff
@@ -65,26 +77,69 @@ func (o *logBufferOptions) override(lookback, blockRate, logLimit uint32) {
 	o.blockRate.Store(blockRate)
 }
 
-func (o *logBufferOptions) windows() int {
-	return int(math.Ceil(float64(o.lookback.Load()) / float64(o.blockRate.Load())))
-}
-
 type logBuffer struct {
+	services.StateMachine
 	lggr logger.Logger
 	opts *logBufferOptions
 	// last block number seen by the buffer
 	lastBlockSeen *atomic.Int64
 	// map of upkeep id to its queue
-	queues map[string]*upkeepLogQueue
-	lock   sync.RWMutex
+	queues          map[string]*upkeepLogQueue
+	lock            sync.RWMutex
+	blockHashes     map[int64]string
+	latestBlockHash atomic.Pointer[common.Hash]
+	history         atomic.Pointer[automation.BlockHistory]
+	subID           int
+	blockHistoryCh  chan automation.BlockHistory
+	threadCtrl      utils.ThreadControl
+	blockSubscriber automation.BlockSubscriber
```

Comment:
q. What happens if the block subscriber has not yet seen the block that enqueue is called for?